src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelopeService.java

```diff
@@ -187,7 +189,15 @@ public void deleteSubmission(SubmissionEnvelope submissionEnvelope, boolean forc
         if (!(submissionEnvelope.isOpen() || forceDelete))
             throw new UnsupportedOperationException("Cannot delete submission if it is already submitted!");
 
-        this.cleanupLinksToSubmissionMetadata(submissionEnvelope);
+        RetryTemplate retry = RetryTemplate.builder()
+                .maxAttempts(5)
+                .fixedBackoff(75)
+                .retryOn(OptimisticLockingFailureException.class)
+                .build();
+        retry.execute(context -> {
+            cleanupLinksToSubmissionMetadata(submissionEnvelope);
```

Comment:
In what scenario does this throw an optimistic lock exception?

Reply:
To be honest, I still can't confirm. It's a race condition that's hard to reproduce. The issue is caused by optimistic locking, and this seems to be the part of the execution that's the most suspect as there's lots of updating happening here. Spefically, the parts where it updates files and biomaterials in succession within the same persistence context could potentially share entities that become out of date the next time they are encountered. 

I applied the retry wholesale for now as it seemed to be the least code to add that doesn't change the behaviour of the code by that much.

Reply:
Yeah I looked at the code in the cleanupLinks() method, looks like it should be safe in a single-threaded context at least. 

I'm guessing it's raised when deleteSubmission() is called twice in succession, maybe by UI users clicking the button multiple times for submissions that take a while to delete. A possible fix might be to block all operations (particularly the delete operation) on the submission completely when deletion is requested and then run the delete() in another thread.